src/rgw/rgw_pubsub.cc

```diff
@@ -462,12 +499,88 @@ std::string rgw_pubsub_dest::to_json_str() const
   return ss.str();
 }
 
+void rgw_pubsub_dest::decode_json(JSONObj* f) {
+  using rgw::notify::DEFAULT_CONFIG;
+  using rgw::notify::DEFAULT_GLOBAL_VALUE;
+  JSONDecoder::decode_json("push_endpoint", push_endpoint, f);
+  JSONDecoder::decode_json("push_endpoint_args", push_endpoint_args, f);
+  JSONDecoder::decode_json("push_endpoint_topic", arn_topic, f);
+  JSONDecoder::decode_json("stored_secret", stored_secret, f);
+  JSONDecoder::decode_json("persistent", persistent, f);
+  std::string ttl;
+  JSONDecoder::decode_json("time_to_live", ttl, f);
+  time_to_live = ttl == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE : std::stoul(ttl);
+
+  std::string max_retry;
+  JSONDecoder::decode_json("max_retries", max_retry, f);
+  max_retries = max_retry == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE
+                                            : std::stoul(max_retry);
+
+  std::string sleep_dur;
+  JSONDecoder::decode_json("retry_sleep_duration", sleep_dur, f);
+  retry_sleep_duration = sleep_dur == DEFAULT_CONFIG ? DEFAULT_GLOBAL_VALUE
+                                                     : std::stoul(sleep_dur);
+}
+
 RGWPubSub::RGWPubSub(rgw::sal::Driver* _driver, const std::string& _tenant)
   : driver(_driver), tenant(_tenant)
 {}
 
-int RGWPubSub::read_topics(const DoutPrefixProvider *dpp, rgw_pubsub_topics& result, 
-    RGWObjVersionTracker *objv_tracker, optional_yield y) const
+int RGWPubSub::get_topics(const DoutPrefixProvider* dpp,
+                          const std::string& start_marker, int max_items,
+                          rgw_pubsub_topics& result, std::string& next_marker,
+                          optional_yield y) const
+{
+  if (!driver->get_zone()->get_zonegroup().supports_feature(
+          rgw::zone_features::notification_v2)) {
+    // v1 returns all topics, ignoring marker/max_items
+    return read_topics_v1(dpp, result, nullptr, y);
+  }
+
+  // TODO: prefix filter on 'tenant:'
+  void* handle = NULL;
+  int ret = driver->meta_list_keys_init(dpp, "topic", start_marker, &handle);
+  if (ret < 0) {
+    return ret;
+  }
+  auto g = make_scope_guard(
+      [this, handle] { driver->meta_list_keys_complete(handle); });
+
+  if (max_items > 1000) {
+    max_items = 1000;
+  }
+  std::list<std::string> topics;
+  bool truncated = false;
+  ret = driver->meta_list_keys_next(dpp, handle, max_items, topics, &truncated);
```

Comment:
i patched this PR and was running the multisite functional test added earlier, and getting `ret = 3` out of `list_keys_next`
here is the debug output
`2024-01-12T12:26:12.080-0500 7fec5cf1d700 20 req 14220179617397124955 0.006999904s sns:pubsub_topics_list RGWRados::pool_iterate: got topic.:topic-dpbmsz-1
2024-01-12T12:26:12.080-0500 7fec5cf1d700 20 req 14220179617397124955 0.006999904s sns:pubsub_topics_list RGWRados::pool_iterate: got topic.:topic-dpbmsz-3
2024-01-12T12:26:12.080-0500 7fec5cf1d700 20 req 14220179617397124955 0.006999904s sns:pubsub_topics_list RGWRados::pool_iterate: got topic.:topic-dpbmsz-2
2024-01-12T12:26:12.081-0500 7fec5cf1d700 20 req 14220179617397124955 0.007999890s sns:pubsub_topics_list successfully got topics
2024-01-12T12:26:12.081-0500 7fec5cf1d700  2 req 14220179617397124955 0.007999890s sns:pubsub_topics_list completing
2024-01-12T12:26:12.082-0500 7fec5cf1d700  2 req 14220179617397124955 0.008999876s sns:pubsub_topics_list op status=3
2024-01-12T12:26:12.082-0500 7fec5cf1d700  2 req 14220179617397124955 0.008999876s sns:pubsub_topics_list http status=500`

Reply:
> i patched this PR and was running the multisite functional test added earlier, and getting `ret = 3` out of `driver->meta_list_keys_next` here is the debug output
> 
> ```
> 2024-01-12T12:26:12.080-0500 7fec5cf1d700 20 req 14220179617397124955 0.006999904s sns:pubsub_topics_list RGWRados::pool_iterate: got topic.:topic-dpbmsz-1
> 
> 2024-01-12T12:26:12.080-0500 7fec5cf1d700 20 req 14220179617397124955 0.006999904s sns:pubsub_topics_list RGWRados::pool_iterate: got topic.:topic-dpbmsz-3
> 
> 2024-01-12T12:26:12.080-0500 7fec5cf1d700 20 req 14220179617397124955 0.006999904s sns:pubsub_topics_list RGWRados::pool_iterate: got topic.:topic-dpbmsz-2
> 
> 2024-01-12T12:26:12.081-0500 7fec5cf1d700 20 req 14220179617397124955 0.007999890s sns:pubsub_topics_list successfully got topics
> 
> 2024-01-12T12:26:12.081-0500 7fec5cf1d700  2 req 14220179617397124955 0.007999890s sns:pubsub_topics_list completing
> 
> 2024-01-12T12:26:12.082-0500 7fec5cf1d700  2 req 14220179617397124955 0.008999876s sns:pubsub_topics_list op status=3
> 
> 2024-01-12T12:26:12.082-0500 7fec5cf1d700  2 req 14220179617397124955 0.008999876s sns:pubsub_topics_list http status=500
> ```
this issue is coz inside the `metadata::lister::next` we are directly returning the `op->next` value which returns the size of the objects retrieved, so we need to explicitly return 0 if there is no error.